### PR TITLE
NanoBSP integration

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Although this was a very nice and welcome exercise, I start to question the actual usefulness of an internal nodes builder. I mean, a map without nodes must strictly be considered in a broken astate nd we'll have to either introduce more checks or relax some others to reliably detect them. So, best we can do is advice map authors to get their levels into a reasonable state before releasing them, and this inevitably includes building nodes. 

This, and the fact that I don't know more than one map that's actually released without nodes.

**Comment by rfomin** [comment]
> I mean, a map without nodes must strictly be considered in a broken state and we'll have to either introduce more checks or relax some others to reliably detect them.

I wonder how many changes would be required?

> This, and the fact that I don't know more than one map that's actually released without nodes.

If we implement DSDA-Doom style UDMF support, we can add limited compatibility with ZDoom. There might be some older maps without nodes. I'm not very familiar with ZDoom maps though.

**Comment by fabiangreffrath** [comment]
If the `SEGS`, `SSECTORS` and `NODES` lump are entirely missing (instead of merely being empty but present), then the `SECTORS` and following lumps shift up to three indices. So, we have to check the presence of some strictly necessary lumps; we have to check if the lump at `ML_SECTORS` is actually called "SECTORS" and react to this; we have to check if the "REJECT" and "BLOCKMAP" lumps do even exist (currently weassume they are and only check their sizes).

https://github.com/fabiangreffrath/woof/blob/3584cb67bb2298ed39fbc4b5b40ad0bdabed6c17/src/doomdata.h#L36-L48

**Comment by fabiangreffrath** [comment]
This is an ugly ad-hoc patch that works with plansifs2.wad:

```patch
diff --git a/src/p_setup.c b/src/p_setup.c
index d319d85..a0fbbce 100644
--- a/src/p_setup.c
+++ b/src/p_setup.c
@@ -274,7 +274,13 @@ void P_LoadSubsectors (int lump)
 void P_LoadSectors (int lump)
 {
   byte *data;
-  int  i;
+  int i = 0;
+
+  while (i < 3 && strcasecmp(lumpinfo[lump].name, "SECTORS"))
+  {
+    lump--;
+    i++;
+  }
 
   numsectors = W_LumpLength (lump) / sizeof(mapsector_t);
   sectors = Z_Malloc (numsectors*sizeof(sector_t),PU_LEVEL,0);
@@ -1157,7 +1163,7 @@ boolean P_LoadBlockMap (int lump)
   // Forces a (re-)building of the BLOCKMAP lumps for loaded maps.
   //
 
-  if (M_CheckParm("-blockmap") || (count = W_LumpLength(lump)/2) >= 0x10000 || count < 4) // [FG] always rebuild too short blockmaps
+  if (M_CheckParm("-blockmap") || lump >= numlumps || strcasecmp(lumpinfo[lump].name, "BLOCKMAP") || (count = W_LumpLength(lump)/2) >= 0x10000 || count < 4) // [FG] always rebuild too short blockmaps
   {
     P_CreateBlockMap();
   }
@@ -1458,7 +1464,7 @@ void P_SegLengths(boolean contrast_only)
 static boolean P_LoadReject(int lumpnum, int totallines)
 {
     int minlength;
-    int lumplen;
+    int lumplen = -1;
     boolean ret;
 
     // Calculate the size that the REJECT lump *should* be.
@@ -1469,7 +1475,8 @@ static boolean P_LoadReject(int lumpnum, int totallines)
     // Otherwise, we need to allocate a buffer of the correct size
     // and pad it with appropriate data.
 
-    lumplen = W_LumpLength(lumpnum);
+    if (lumpnum < numlumps && !strcasecmp(lumpinfo[lumpnum].name, "REJECT"))
+      lumplen = W_LumpLength(lumpnum);
 
     if (lumplen >= minlength)
     {
```

**Comment by rfomin** [comment]
> This is an ugly ad-hoc patch that works with plansifs2.wad

Maybe only planisf2.wad should be loaded for now? Perhaps we will found some maps later.

**Comment by fabiangreffrath** [comment]
Alright, got me. I'll keep in working on this.

File: src/p_setup.c [hunk 1]
```diff
@@ -275,6 +276,12 @@ void P_LoadSectors (int lump)
   byte *data;
   int  i;
 
+  // [FG] SEGS, SSECTORS, NODES lumps missing?
+  for (i = ML_SECTORS; !W_LumpExistsWithName(lump, "SECTORS") && i > ML_VERTEXES; i--)
+  {
+    lump--;
+  }
+
```
Comment (fabiangreffrath) [inline src/p_setup.c:284]: This needs an error message if no lump with that name is found.

**Comment by fabiangreffrath** [comment]
Should be ready now.

**Review by rfomin — APPROVED** [review]
LGTM, planisf2.wad works fine.

**Pull request merged**